### PR TITLE
shared.lua: add tbl_deep_extend

### PR DESCRIPTION
**Pull request by h-michael**
ref: https://github.com/neovim/nvim-lsp/pull/98

### CODE REVIEW

**Comment by tjdevries** [comment]
@teto LGTM, any concerns?

**Comment by teto** [comment]
seems like `vim.tbl_extend` could be redefined in terms of this one ?

**Comment by tjdevries** [comment]
Yeah, @h-michael can you refactor it so that `tbl_extend` and `tbl_deep_extend` use the definitions internally. Should be able to be basically the same, except they just do something different when setting the values.

**Comment by h-michael** [comment]
@teto @tjdevries I refactor common logic :)

File: runtime/lua/vim/shared.lua [hunk 1]
```diff
@@ -228,7 +219,9 @@ function vim.tbl_extend(behavior, ...)
     vim.validate{["after the second argument"] = {tbl,'t'}}
     if tbl then
       for k, v in pairs(tbl) do
-        if behavior ~= 'force' and ret[k] ~= nil then
+        if type(v) == 'table' and deep_extend and not vim.tbl_islist(v) then
```
Comment (tjdevries) [inline runtime/lua/vim/shared.lua:222]: Why `and not vim.tbl_islist(v)`?

I would expect deep extend to also copy table elements?

```lua
points = { { x = 1, y = 1}, { x = 3, y = 5} }
```

Comment (tjdevries) [reply runtime/lua/vim/shared.lua:222]: Oh, I see -- it doesn't work for non-map-like tables. Why is that?

Comment (tjdevries) [reply runtime/lua/vim/shared.lua:222]: Actually, nvm, this doesn't apply.

**Pull request merged**